src/sql/parts/notebook/notebookInput.ts

```diff
@@ -18,28 +18,88 @@ import { IStandardKernelWithProvider } from 'sql/parts/notebook/notebookUtils';
 import { IDialogService } from 'vs/platform/dialogs/common/dialogs';
 import Severity from 'vs/base/common/severity';
+import { IInstantiationService } from 'vs/platform/instantiation/common/instantiation';
+import { ISaveOptions } from 'vs/workbench/services/textfile/common/textfiles';
+import { ITextModelService } from 'vs/editor/common/services/resolverService';
+import { INotebookModel } from 'sql/parts/notebook/models/modelInterfaces';
+import { TextFileEditorModel } from 'vs/workbench/services/textfile/common/textFileEditorModel';
 
 export type ModeViewSaveHandler = (handle: number) => Thenable<boolean>;
 
 
-export class NotebookInputModel extends EditorModel {
+export class NotebookEditorModel extends EditorModel {
 	private dirty: boolean;
 	private readonly _onDidChangeDirty: Emitter<void> = this._register(new Emitter<void>());
 	private _providerId: string;
+	private _providers: string[];
+	private _connectionProfileId: string;
 	private _standardKernels: IStandardKernelWithProvider[];
 	private _defaultKernel: azdata.nb.IKernelSpec;
 	constructor(public readonly notebookUri: URI,
-		private readonly handle: number,
 		private _isTrusted: boolean = false,
-		private saveHandler?: ModeViewSaveHandler,
-		provider?: string,
-		private _providers?: string[],
-		private _connectionProfileId?: string) {
-
+		private textEditorModel: TextFileEditorModel,
+		@INotebookService private notebookService: INotebookService
+	) {
 		super();
-		this.dirty = false;
-		this._providerId = provider;
 		this._standardKernels = [];
+		this._register(this.notebookService.onNotebookEditorAdd(notebook => {
+			if (notebook.id === this.notebookUri.toString()) {
+				// Hook to content change events
+				this._register(notebook.model.contentChanged(e => this.updateModel()));
+				this._register(notebook.model.kernelChanged(e => this.updateModel()));
+			}
+		}));
+	}
+
+	public get contentString(): string {
+		//TODO: Handle Save As
+		// if(!this.textEditorModel || !this.textEditorModel.textEditorModel){
+		// 	return '';
+		// }
+		let model = this.textEditorModel.textEditorModel;
+		return model.getValue();
+	}
+
+	save(options: ISaveOptions): TPromise<void> {
+		options.force = false;
+		this.updateModel();
+		return this.textEditorModel.save(options);
+	}
+
+	isDirty(): boolean {
+		//TODO: Need to handle dirty with file save
+		// if(!this.textEditorModel){
+		// 	return true;
+		// }
+		return this.textEditorModel.isDirty();
+	}
+
+	public setDirty(dirty: boolean): void {
+		if (this.dirty === dirty) {
+			return;
+		}
+		this.dirty = dirty;
+		this._onDidChangeDirty.fire();
+	}
+
+	private updateModel(): void {
+		let notebookModel = this.getNotebookModel();
+		if (notebookModel && this.textEditorModel && this.textEditorModel.textEditorModel) {
+			let content = JSON.stringify(notebookModel.toJSON(), undefined, '    ');
+			this.textEditorModel.textEditorModel.setValue(content);
```

Comment:
Are there plans to enable undo/redo? this won't achieve those features.

Reply:
Changed it to applyEdits. Need to test all the scenarios.